File: electro-server/src/main/java/com/electro/entity/promotion/Promotion.java
package com.electro.entity.promotion;

import com.electro.entity.BaseEntity;
import com.electro.entity.product.Product;
import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import lombok.experimental.Accessors;

import javax.persistence.CascadeType;
import javax.persistence.Column;
import javax.persistence.Entity;
import javax.persistence.JoinColumn;
import javax.persistence.JoinTable;
import javax.persistence.ManyToMany;
import javax.persistence.Table;
import java.time.Instant;
import java.util.ArrayList;
import java.util.List;

@AllArgsConstructor
@NoArgsConstructor
@Getter
@Setter
@Accessors(chain = true)
@Entity
@Table(name = "promotion")
public class Promotion extends BaseEntity {

    @Column(name = "name", nullable = false)
    private String name;

    @Column(name = "start_date", nullable = false)
    private Instant startDate;

    @Column(name = "end_date", nullable = false)
    private Instant endDate;

    @Column(name = "percent", nullable = false)
    private Integer percent;

    @Column(name = "status", nullable = false, columnDefinition = "TINYINT")
    private Integer status;

    @ManyToMany(cascade = {CascadeType.MERGE})
    @JoinTable(
            name = "product_promotion",
            joinColumns = @JoinColumn(name = "promotion_id", nullable = false),
            inverseJoinColumns = @JoinColumn(name = "product_id", nullable = false)
    )
    private List<Product> products = new ArrayList<>();
}

File: electro-client/src/models/Promotion.ts
import BaseResponse from 'models/BaseResponse';
import { ProductResponse } from 'models/Product';

export interface PromotionResponse extends BaseResponse {
  name: string;
  startDate: string;
  endDate: string;
  percent: number;
  status: number;
  products: ProductResponse[];
}

export interface PromotionRequest {
  name: string;
  startDate: string;
  endDate: string;
  percent: number;
  status: number;
  productIds: number[];
  categoryIds: number[];
}

export interface PromotionCheckingResponse {
  promotionable: boolean;
}
